**Clustering Illumina MiSeq paired-end reads of MHC sequences**

Clustering of paired-end reads is conducted using the script `mhc_clustering.py`.

```
mhc_clustering.py -h
```

The subfolder `lib` contains required functions and may contain hidden marko models created from multiple sequence alignment of published mhc alleles using **HMMER3**.  


*This repository was forked from https://github.com/nextgenusfs/mhc_cluster.git` and subsequently modified. See details there.*


**Dependent software**

* python,
* biopython
* USEARCH8 (http://www.drive5.com/usearch)
* VSEARCH (https://github.com/torognes/vsearch)
* HMMER3 v3.1b2



